src/parcsr_ls/par_amg.c

```diff
@@ -316,6 +316,11 @@ hypre_BoomerAMGCreate( void )
 
    amg_data = hypre_CTAlloc(hypre_ParAMGData, 1, HYPRE_MEMORY_HOST);
 
+   /* Set base solver function pointers */
+   hypre_ParAMGDataBaseSetup(amg_data)   = (HYPRE_PtrToSolverFcn)  HYPRE_BoomerAMGSetup;
```

Comment:
Hi @victorapm .  Consider using the casting idea in Tzanio's note.  Since BoomerAMG "is a" solver, you should be allowed to cast it to that base type.  Your implementation of `hypre_Solver` and `hypre_ParAMGData` allows for that.  It would save needing to add accessor macros for every solver, too.

This was the big thing I wanted to be sure was portable, and it apparently is in C11 (see the first comment in the Roundup issue 305).  I can try to check on C99 since that's our current minimum standard.

Reply:
It looks like this is also C99 compliant.  See item 13 in section 6.7.2.1 here:
  https://www.dii.uchile.cl/~daespino/files/Iso_C_1999_definition.pdf

Reply:
Hi Rob, I read it but it's not clear to me what should be changed here. Can you share a code snippet of what you envision?

Reply:
Also, feel free to implement any changes directly to this branch. Maybe what you have is simple and I just can't see it :)

Reply:
Can you add `base.h` to `utilities`?  I can give this a go once it's there.  In the meantime, the code snippet in Tzanio's note is this:
```
   hypre_ParAMGData *amg_data;
   hypre_Solver *solver;

   amg_data = hypre_CTAlloc(hypre_ParAMGData, 1);

   /* Construct the base class */
   solver = (hypre_Solver *) amg_data;
   solver -> use_as_preconditioner = 0;
   solver -> Setup  = (HYPRE_PtrToSolverFcn) HYPRE_BoomerAMGSetup;
   solver -> Solve  = (HYPRE_PtrToSolverFcn) HYPRE_BoomerAMGSolve;
   solver -> SolveT = (HYPRE_PtrToSolverFcn) HYPRE_BoomerAMGSolveT;

   /* Fill the rest of amg_data */
   /* ... */
```
Notice the cast in the `solver =` line.  Both `amg_data` and `solver` point to the same thing.  Note that `hypre_ParAMGData` is really a `hypre_ParAMGSolver` now, so that could also be renamed at some point.

Reply:
Hi Rob, `base.h` is in `_hypre_utilities.h`. Do you mean adding it under `HYPRE_utilities.h`?

Reply:
Sorry, now I understand it. Just added the missing file with [71d546c](https://github.com/hypre-space/hypre/pull/980/commits/71d546c015e1ee561401aa15f83f0ed59b317e55)

Thanks for helping with the PR, let me know if I can something else

Reply:
I implemented your suggestions in https://github.com/hypre-space/hypre/pull/980/commits/9192564c6b7c7ed74b88cbcda3c5a6315d474767

Please, feel free to adjust. This looks better than before, thanks!